src/analysis/retail/shaman/elemental/modules/features/SpenderWindow.tsx

```diff
@@ -0,0 +1,295 @@
+import ThresholdPerformancePercentage, {
+  LTEThreshold,
+  GTEThreshold,
+} from './shared/ThresholdPerformancePercentage';
+import SPELLS from 'common/SPELLS';
+import TALENTS from 'common/TALENTS/shaman';
+import { Talent } from 'common/TALENTS/types';
+import { formatDuration } from 'common/format';
+import { SpellIcon, SpellLink, TooltipElement } from 'interface';
+import { explanationAndDataSubsection } from 'interface/guide/components/ExplanationRow';
+import Analyzer, { Options, SELECTED_PLAYER } from 'parser/core/Analyzer';
+import Events, { CastEvent } from 'parser/core/Events';
+import Enemies from 'parser/shared/modules/Enemies';
+import { ON_CAST_BUFF_REMOVAL_GRACE_MS } from '../../constants';
+
+interface ActiveSpenderWindow {
+  timestamp: number;
+  motePresent: boolean;
+  flameshockPresent: boolean;
+}
+
+interface FinishedSpenderWindow extends ActiveSpenderWindow {
+  elshocksPresent: boolean;
+  sopUse: CastEvent;
+}
+
+const SOP_SPENDERS = [
+  SPELLS.LIGHTNING_BOLT,
+  TALENTS.CHAIN_LIGHTNING_TALENT,
+  SPELLS.FLAME_SHOCK,
+  TALENTS.FROST_SHOCK_TALENT,
+  TALENTS.LAVA_BURST_TALENT,
+];
+
+const GOOD_SOP_SPENDERS = [SPELLS.LIGHTNING_BOLT.id, TALENTS.CHAIN_LIGHTNING_TALENT.id];
+
+const PERFECT_WINDOWS_THRESHOLD: GTEThreshold = {
+  type: 'gte',
+  perfect: 0.9,
+  good: 0.8,
+  ok: 0.7,
+};
+const WRONG_SOP_THRESHOLD: LTEThreshold = {
+  type: 'lte',
+  perfect: 0,
+  good: 0.1,
+  ok: 0.15,
+};
+const MISSING_ELSHOCKS_THRESHOLD: LTEThreshold = {
+  type: 'lte',
+  perfect: 0,
+  good: 0.2,
+  ok: 0.3,
+};
+const MISSING_MOTE_THRESHOLD: LTEThreshold = {
+  type: 'lte',
+  perfect: 0,
+  good: 0.2,
+  ok: 0.3,
+};
+const MISSING_FLAMESHOCK_THRESHOLD: LTEThreshold = {
+  type: 'lte',
+  perfect: 0,
+  good: 0.2,
+  ok: 0.3,
+};
+
+class SpenderWindow extends Analyzer {
+  static dependencies = {
+    enemies: Enemies,
+  };
+
+  enabled: boolean;
+
+  protected enemies!: Enemies;
+
+  activeSpenderWindow: ActiveSpenderWindow | null;
+  spenderWindows: FinishedSpenderWindow[];
+
+  private stSpender: Talent = TALENTS.EARTH_SHOCK_TALENT;
+
+  constructor(options: Options) {
+    super(options);
+
+    this.activeSpenderWindow = null;
+    this.spenderWindows = [];
+
+    if (this.selectedCombatant.hasTalent(TALENTS.ELEMENTAL_BLAST_TALENT)) {
+      this.stSpender = TALENTS.ELEMENTAL_BLAST_TALENT;
+    }
+
+    this.enabled = this.selectedCombatant.hasTalent(TALENTS.SURGE_OF_POWER_TALENT);
+
+    /* There is no point in tracking this at all if the player doesn't have SoP */
+    if (!this.enabled) {
+      return;
+    }
+
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(this.stSpender), this.onSpender);
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(SOP_SPENDERS), this.onSopSpender);
+  }
+
+  onSpender(event: CastEvent) {
+    if (this.activeSpenderWindow) {
+      return;
+    }
+
+    this.activeSpenderWindow = {
```

Comment:
[Nit] There's another way to write this with the null-coalescing operator.  Not sure if you like that more or less.

```typescript
this.activeSpenderWindow = this.activeSpenderWindow ?? {
  ...
};
```

Reply:
I found that in another PR of this repo, but I decided against using it as AFAIK it isn't really a common operator outside TS. I tend to like to be explicit, so the code reads "If there is an active window, don't do anything in this function". I feel '??=' requires some interpretation to reach my intent.

Can of course use it if it is an established pattern in TS.